misc/determine-version.py

```diff
@@ -179,8 +196,11 @@ def version_cmp(a, b):
     git_rev = subprocess.Popen(["git", "rev-parse", tag + "^{}"],
                                stdout=subprocess.PIPE)
     rev = git_rev.stdout.readlines()[0].strip()
-    if rev == recent_rev:
-        # Ignore tags that point at the same commit.
+    if not in_master_branch and rev == recent_rev:
+        # Ignore tags that point at the same commit as the most recent
+        # tag, i.e. ignore the same tag (TODO COMMENT WHY?). But if we
+        # are on master branch, we need that exact tag, as it might be
+        # the most recent ever issued.
```

Comment:
You have to ignore tags that point to the same commit because otherwise the two cases "Z with 3.9, and M with 3.7" in the logic description at the top are indistinguishable, and you want to have different behavior for the two cases (increase patch version on Z, increase minor version on M).

Maybe you should update that comment btw, since you are changing the logic slightly.